src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java

```diff
@@ -652,28 +649,62 @@ private static Optional<DetailAST> findCommentUnder(DetailAST packageDef) {
-     * @return {@code true} if found any blank line within the range, {@code false}
+     * @param currentEnd the last token of the current subtree
+     * @param nextBegin the first token of the next subtree
+     * @return {@code true} if found any blank line, {@code false}
      *         otherwise
      */
-    // suppress deprecation until https://github.com/checkstyle/checkstyle/issues/11166
-    @SuppressWarnings("deprecation")
-    private boolean hasEmptyLine(int startLine, int endLine) {
-        // Initial value is false - blank line not found
-        boolean result = false;
-        final FileContents fileContents = getFileContents();
-        for (int line = startLine; line <= endLine; line++) {
-            // Check, if the line is blank. Lines are numbered from 0, so subtract 1
-            if (fileContents.lineIsBlank(line - 1)) {
-                result = true;
-                break;
+    private boolean hasEmptyLine(DetailAST currentEnd, DetailAST nextBegin) {
+        final int currentEndLineNo = currentEnd.getLineNo();
+        final int nextBeginLineNo = nextBegin.getLineNo();
+        final int linesBetween = nextBeginLineNo - currentEndLineNo - 1;
+
+        return getCommentsLineCount(nextBegin, currentEnd) < linesBetween;
+    }
+
+    /**
+     * Get the number of commented lines between the given token and the previous token.
+     *
+     * @param currentToken the current token
+     * @param previousToken the previous token
+     * @return the number of comment lines between the current token and the previous token
+     */
+    private int getCommentsLineCount(DetailAST currentToken, DetailAST previousToken) {
+        int commentsLineCount = 0;
+        DetailAST commentAst = findFirstCommentToken(currentToken);
+
+        while (commentAst != null && TokenUtil.isCommentType(commentAst.getType())) {
+            if (TokenUtil.areOnSameLine(commentAst, previousToken)
+                    || TokenUtil.areOnSameLine(commentAst, currentToken)) {
+                commentAst = commentAst.getNextSibling();
+                continue;
             }
+            final DetailAST commentEnd = commentAst.getLastChild();
+            commentsLineCount += commentEnd.getLineNo() - commentAst.getLineNo() + 1;
+            commentAst = commentAst.getNextSibling();
         }
-        return result;
+        return commentsLineCount;
+    }
+
+    /**
+     * Finds the first comment token in the subtree of the given token.
+     *
+     * @param ast token to start the search
+     * @return the first comment token, or {@code null} if not found
+     */
+    private DetailAST findFirstCommentToken(DetailAST ast) {
+        final DetailAST modifersAst = ast.findFirstToken(TokenTypes.MODIFIERS);
+        final DetailAST commentAst;
+
+        if (modifersAst != null) {
+            commentAst = modifersAst.getFirstChild();
+        }
+        else {
+            commentAst = ast.getFirstChild();
+        }
+        return commentAst;
```

Comment:
I will work more on this method. I believe that this will cause NPE. more cases need to be handled to get the first comment token